anonlink/serialization.py

```diff
@@ -0,0 +1,221 @@
+_ARRAY_FLOAT_TYPES = 'df'
+_ARRAY_UINT_LEN_TO_FMT = {_array.array(t).itemsize: t
+                          for t in _ARRAY_UINT_TYPES}
+_ARRAY_FLOAT_LEN_TO_FMT = {_array.array(t).itemsize: t
+                           for t in _ARRAY_FLOAT_TYPES}
+
+_HEADER_STRUCT = _struct.Struct('<BBBB')
+
+
+_CandidatePair = _typing.Tuple[float, int, int, int, int]
+_CandidatePairIter = _typing.Iterable[_CandidatePair]
+
+# This is subject to change.
+# The format is composed of a header and a sequence of entries.
+# The header is composed of:  version (1 byte), number of bytes in
+# similarity (1 byte), number of bytes in dataset index (1 byte), number
+# of bytes in record index (1 byte).
+# Every entry is composed of a similarity as a IEEE floating point
+# value, a two datset indices, and two record indices. Their sizes as
+# determined by the header.
+# All integers are unsigned. All values are *little-endian* (yes, this
+# breaks convention). The number of candidate matchings is not included
+# on purpose: if needed, it can be computed from the length of the file.
+
+
+def _entry_struct(
+    sim_t_size: int,
+    dset_i_t_size: int,
+    rec_i_t_size: int
+) -> _struct.Struct:
+    try:
+        sim_t = _STRUCT_FLOAT_LEN_TO_FMT[sim_t_size]
+    except KeyError:
+        msg = f'floats of {sim_t_size} bytes are not supported'
+        raise ValueError(msg) from None
+    try:
+        dset_i_t = _STRUCT_UINT_LEN_TO_FMT[dset_i_t_size]
+    except KeyError:
+        msg = f'indices of {dset_i_t_size} bytes are not supported'
+        raise ValueError(msg) from None
+    try:
+        rec_i_t = _STRUCT_UINT_LEN_TO_FMT[rec_i_t_size]
+    except KeyError:
+        msg = f'indices of {rec_i_t_size} bytes are not supported'
+        raise ValueError(msg) from None
+    return _struct.Struct(f"<{sim_t}2{dset_i_t}2{rec_i_t}")
+    
+
+def _dump_from_iterable(
+    f: _typing.BinaryIO,
+    sim_t_size: int,
+    dset_i_t_size: int,
+    rec_i_t_size: int,
+    iterable: _CandidatePairIter
+) -> None:
+    # Fail without writing if the parameters are incorrect.
+    entry_struct = _entry_struct(
+        sim_t_size, dset_i_t_size, rec_i_t_size)
+    
+    # Write header.
+    f.write(_HEADER_STRUCT.pack(1, sim_t_size, dset_i_t_size, rec_i_t_size))
+    
+    for entry in iterable:
+        f.write(entry_struct.pack(*entry))
+
+
+def _entries_iterable(
+    f: _typing.BinaryIO,
+    entry_struct: _struct.Struct
+) -> _CandidatePairIter:
+    buffer = f.read(entry_struct.size)
+    while buffer:
+        if len(buffer) != entry_struct.size:
+            raise ValueError('ran out of input')
+        yield _typing.cast(_CandidatePair, entry_struct.unpack(buffer))
+        buffer = f.read(entry_struct.size)
+
+
+def _load_to_iterable(
+    f: _typing.BinaryIO
+) -> _typing.Tuple[_CandidatePairIter, int, int, int]:
+    # Make sure that f.read(n) returns exactly n bytes.
+    if isinstance(f, _io.RawIOBase):
+        f = _io.BufferedWriter(f)
```

Comment:
Usually in Python we would `try` and except a runtime failure, instead of checking types at runtime. Why not here?

Reply:
The point of this section is to ensure that `f.read(n)` will return exactly `n` bytes. This is not guaranteed in a RawIOBase instance but is guaranteed in a BufferedIOBase instance. Raw IO can be easily turned into Buffered IO.

There is no good way to test whether `f.read(n)` will always return `n` bytes, aside from checking whether f is an instance of BufferedIOBase. For example, I can’t just call `f.read(5)` and see if 5 bytes are returned: the call could return 5 bytes by chance even if it’s not buffered, and the file would be advanced. Type checking hence appears to be the best way to do this.

Further, I argue that this does not violate Python’s principle of duck typing. In duck typing, we want a class A that implements the interface of class B to pass for the latter whenever is is needed (without having to subclass). Since BufferedIOBase and RawIOBase are metaclasses, the classes implementing those interfaces can simply .register() itself to be recognised as a ’subclass’. In fact, any sane implementation should do this.

Reply:
Agree